helper/dhutil/dhutil.go

```diff
@@ -36,21 +39,47 @@ func GeneratePublicPrivateKey() ([]byte, []byte, error) {
 
 // generateSharedKey uses the private key and the other party's public key to
 // generate the shared secret.
-func GenerateSharedKey(ourPrivate, theirPublic []byte) ([]byte, error) {
+func GenerateSharedSecret(ourPrivate, theirPublic []byte) ([]byte, error) {
 	if len(ourPrivate) != 32 {
 		return nil, fmt.Errorf("invalid private key length: %d", len(ourPrivate))
 	}
 	if len(theirPublic) != 32 {
 		return nil, fmt.Errorf("invalid public key length: %d", len(theirPublic))
 	}
 
-	var scalar, pub, secret [32]byte
-	copy(scalar[:], ourPrivate)
-	copy(pub[:], theirPublic)
+	return curve25519.X25519(ourPrivate, theirPublic)
+}
+
+// DeriveSharedKey uses HKDF to derive a key from a shared secret and public keys
+func DeriveSharedKey(secret, ourPublic, theirPublic []byte) ([]byte, error) {
+	// Derive the final key from the HKDF of the secret and public keys.
+
+	//These must be consistently ordered we're on "our" side of key negotiation or the other.
```

Comment:
Can you explain a bit more (or link to some reading material) why the result of the comparison determines the order of arguments?

Reply:
Internally, HKDF hashes the secret and two public keys.  If Alice and Bob are doing DH key exchange, Alice calculates:

HKDF(secret, A, B)  since ourPublic is A.

Bob calculates HKDF(secret, B, A), since Bob's ours is B.  That produces a different value.  Now we only care that both public keys participate in the derivation, so simply sorting them so they are in a consistent numerical order (either one would do) arrives at an agreed value.

Reply:
Mind adding that as a comment?